src/niquests/middlewares.py

```diff
@@ -0,0 +1,102 @@
+import typing
+from inspect import iscoroutinefunction
+from abc import ABC
+
+if typing.TYPE_CHECKING:
+    from .models import PreparedRequest, Response
+
+
+class Middleware(ABC):
+    """Base class for synchronous request lifecycle middlewares."""
+
+    def on_request(self, request: 'PreparedRequest', *args, **kwargs) -> None:
+        """Called after the request is prepared but before it’s sent.
+
+        :param request: The prepared request object.
+        :param kwargs: Additional context.
+        """
+        pass
+
+    def on_response(self, response: 'Response', *args, **kwargs) -> None:
+        """Called after receiving a response.
+
+        :param response: The response object.
+        :param kwargs: Additional context.
+        """
+        pass
+
+
+class MiddlewareExecutor:
```

Comment:
for performance sake, maybe do that, but without a class executor. it seems we can just reach for on_response in each middleware. provided we have AsyncMiddleware(ABC) and Middleware(ABC) (that would remove the need for iscoroutine)